query-planner-js/src/buildPlan.ts

```diff
@@ -4376,44 +4383,23 @@ function computeGroupsForTree(
             );
             updated.group = requireResult.group;
             updated.path = requireResult.path;
-
-            // add __typename to site where we are retrieving context from
-            // this is necessary because the context rewrites path will start with a type condition
-            if (contextToSelection) {
-              assert(isCompositeType(edge.head.type), () => `Expected a composite type for ${edge.head.type}`);
-              updated.group.addAtPath(path.inGroup().concat(new Field(edge.head.type.typenameField()!)));
-            }
-
+            
             if (contextToSelection) {
-              const newContextToConditionsGroups = new Map<string, FetchGroup[]>([...contextToConditionsGroups]);
-              for (const context of contextToSelection) {
-                newContextToConditionsGroups.set(context, [group]);
+              if (requireResult.createdGroups.length === 0) {
+                assert(isCompositeType(edge.head.type), () => `Expected a composite type for ${edge.head.type}`);
+                updated.group.addAtPath(path.inGroup().concat(new Field(edge.head.type.typenameField()!)));
               }
-              updated.contextToConditionsGroups = newContextToConditionsGroups;
-            }
-            updateCreatedGroups(createdGroups, ...requireResult.createdGroups);
-          } else if (conditions) {
-            // add __typename to site where we are retrieving context from
-            // this is necessary because the context rewrites path will start with a type condition
-            assert(isCompositeType(edge.head.type), () => `Expected a composite type for ${edge.head.type}`);
-            group.addAtPath(path.inGroup().concat(new Field(edge.head.type.typenameField()!)));
-
-            const conditionsGroups = computeGroupsForTree({
-              dependencyGraph, 
-              pathTree: conditions, 
-              startGroup: group, 
-              initialGroupPath: path, 
-              initialDeferContext: deferContextForConditions(deferContext),
-            });
-
-            if (contextToSelection) {
               const newContextToConditionsGroups = new Map<string, FetchGroup[]>([...contextToConditionsGroups]);
-              for (const context of contextToSelection) {
-                newContextToConditionsGroups.set(context, [group, ...conditionsGroups]);
+              // If a new group is created for the conditions, make sure that we add that as a dependency
+              // so that using a context will not occur within the same fetch group as it is retrieved
+              if (!requireResult.newGroupIsUnneeded || group.isTopLevel) {
```

Comment:
Not sure why we'd only set the condition groups when this boolean expression is true, in general we need these condition groups so the downstream `@fromContext` will be able to depend on these.